anti_patterns/borrow_clone.md

```diff
@@ -0,0 +1,55 @@
+# Clone to satisfy the borrow checker
+
+## Description
+
+The borrow checker prevents Rust users from developing otherwise unsafe code by
+ensuring that either: only one mutable reference exists, or potentially many but
+all immutable references exist. If the code written does not hold true to these 
+conditions, this anti-pattern arises when the developer resolves the compiler
+error by cloning the variable.
+
+
+## Example
+
+```rust
+// define any variable
+let mut x = 5;
+
+// Borrow `x` -- but clone it first
+let y = &mut (x.clone()); 
+
+// perform some action on the borrow to prevent rust from optimizing this
+//out of existence
+*y += 1;
+
+// without the x.clone() two lines prior, this line would fail on compile as
+// x has been borrowed
+// thanks to x.clone(), x was never borrowed, and this line will run.
+println!("{}", x);
+```
+
+
+## Motivation
+
+It is tempting, particularly for beginners, to use this pattern to resolve
+confusing issues with the borrow checker. However, there are serious
+consequences. Using .clone() causes a copy of the data to be made. Any changes
+between the two are not synchronized -- as if two completely separate variables
+exist.
+
+There are special cases -- `Rc<T>` is designed to handle clones intelligently.
+It internally manages exactly one copy of the data, and cloning it will only
+clone the reference.
+
+In general, clones should be deliberate, with full understanding of the
+consequences. If a clone is used to make a borrow checker error disappear,
+that's a good indication this anti-pattern may be in use.
+
+If an unnecessary clone is suspected, The Rust Book's chapter on Ownership
+should be understood fully before assessing whether the clone is required or not.
+
```

Comment:
```suggestion

Even though `clone` is an indication of a bad pattern. Sometimes, **it is fine to write write inefficient code**, in cases such as when the developer is still new to ownership or when code needs to be written in a hurry like hackathon or prototype.
```
We should have a big note that **it is fine to write inefficient code** to reduce the developers (or beginners) OCD on clones when coming to rust, these may ease their start.

Maybe we should also mention that `clone` is fine for `struct`, it is not always necessary to have a `struct` to use borrowed types, especially on cases where the data needs to be sent across threads or processed in a special way. Putting the original owned value may be easier. Not sure if this is a good place to talk about this, this is something not taught in the book but written online mostly.

Reply:
Perhaps also mention running clippy on the code for [its](https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone) many, [many](https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_copy) clone [lints](https://rust-lang.github.io/rust-clippy/master/index.html#map_clone) they'd [hit](https://rust-lang.github.io/rust-clippy/master/index.html#clone_double_ref)?

Reply:
jhwgh, I addressed this in the comment below.
pickfire, I will rephrase this like this:

```
Even though `.clone()` is an indication of a bad pattern, sometimes
**it is fine to write inefficient code**, in cases such as when:
- the developer is still new to ownership
- the code doesn't have great speed or memory constraints
  (like hackathon projects or prototypes)
- satisfying the borrow checker is really complicated and you prefer to
  optimize readability over performance
```

If you like it you can edit your comment (also partially) :smile:

Reply:
Yes, that is better.

Reply:
Feel free to fix it and edit your comment as you prefer :)